File: scripts/ExtendSentiWords.java
import java.io.BufferedInputStream;
import java.io.InputStreamReader;
import java.io.*;
import java.io.DataInputStream;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.File;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Iterator;
import java.util.List;
import java.util.Map.Entry;
import java.util.Set;
import java.util.TreeSet;

public class ExtendSentiWords{
	
	public static void maini(String[] args) throws IOException {
		Word2VEC vec = new Word2VEC();
		vec.loadModel("/Users/wdxu/git/qa-demo/data/word2vec/lookup.bin");
		/*
		System.out.println("One word analysis");
	
		System.out.println("*******************************");
		System.out.println("Three word analysis");
		result = vec.analogy("中华民国", "中华人民共和国", "毛泽东");
		iter = result.iterator();
		while (iter.hasNext()) {
			Word2VEC.WordEntry word = (Word2VEC.WordEntry) iter.next();
			System.out.println(word.name + " " + word.score);
		}
		*/
	}

	public static void main(String[] args) throws IOException{
		
		Word2VEC vec = new Word2VEC();
		vec.loadModel("/Users/wdxu/git/qa-demo/data/word2vec/lookup.bin");
		System.out.println("One word analysis");
        try {
                String encoding="UTF-8";
                File infile = new File("../data/senti-words/mutals-and-adj");
				File outfile = new File("../data/senti-words/extend-mutals-and-adj");
				BufferedWriter output = new BufferedWriter(new FileWriter(outfile));
			
				InputStreamReader read = new InputStreamReader(new FileInputStream(infile),encoding);
                BufferedReader bufferedReader = new BufferedReader(read);
                String lineTxt = null;
                while((lineTxt = bufferedReader.readLine()) != null){
					String[] words = lineTxt.split("\t");
					System.out.print(words[0] + " ");
					output.write(words[0] + ' ');
					Set<Word2VEC.WordEntry> result = find_near(vec, words[0]);
					if( result != null){
						Iterator iter = result.iterator();
						while (iter.hasNext()) {
							Word2VEC.WordEntry word = (Word2VEC.WordEntry) iter.next();
							output.write(word.name + " " + word.score + "\t");
						}
					}
					output.write("\n");
				}
                read.close();

		} catch (Exception e) {
            System.out.println("读取文件内容出错");
            e.printStackTrace();
        }
		
	}

	public static Set<Word2VEC.WordEntry> find_near(Word2VEC vec, String org){
		Set<Word2VEC.WordEntry> result = new TreeSet<Word2VEC.WordEntry>();
		try{
			result = vec.distance(org);
			Iterator iter = result.iterator();
			if (iter.hasNext()) {
				Word2VEC.WordEntry word = (Word2VEC.WordEntry) iter.next();
				System.out.println(word.name + " " + word.score);
			}
			return result;
		} catch( Exception e){
			System.out.println("Not found");
			return null;
		}
	}

}

File: scripts/extend_senti_words.py
#!/usr/bin/python

import codecs
import numpy as np

def extend_senti_words(ifn, wfn, ofn):
	sf = codecs.open(ifn, 'r', 'utf-8')
	wf = codecs.open(wfn, 'r', 'utf-8')
	of = codecs.open(ofn, 'w', 'utf-8')

	word_vectors = {}
	line = wf.readline()
	while line:
		words = line.split()
		if not len(words):
			print words
			line = wf.readline()
			continue
		word_vectors[words[0]] = []
		for i in range(1, len(words)):
			word_vectors[words[0]].append(float(words[i]))
		line = wf.readline()

	word_abs_dis = {}
	for key in word_vectors.keys():
		len_1 = 0
		vec = word_vectors[key]
		for x in vec:
			len_1 += x*x
		word_abs_dis[key] = vec[0] * 1 / np.sqrt(len_1)

	line = sf.readline()
	while line:
		word = line.split()[0]
		nears = find_near(word, word_vectors, 1)
		print nears
		line = sf.readline
	
	return

def find_near(word, word_vectors, cnt):
	if word not in word_vectors.keys():
		return None

	word_vector = word_vectors[word]
	dis = {}

	for key in word_vectors.keys():
		dis[w] = cal_cos_dis(word_vectors[key], word_vectors[word])
	
	sorted_dis = sorted(dis, key = lambda d: d[1], reverse = True)
	
	return sorted_dis[0: 5]

def cal_cos_dis(w1, w2):
	len_1 = 0
	len_2 = 0
	
	for x in w1:
		len_1 += x*x
	for x in w2:
		len_2 += x*x
	
	dis = 0
	for i in range(len(w1)):
		dis += w1[i] * w2[i]

	return dis/np.sqrt(len_1)/np.sqrt(len_2)

extend_senti_words('../data/senti-words/mutals-and-adj', '/Users/wdxu/git/qa-demo/data/word2vec/lookup.txt', '../data/senti-words/extend-mutals-and-adj')
